### DISCUSSION TITLE
Good replacement for prev_execution_date template field in Airflow 3

### DISCUSSION THREAD

**Discussion by k-tomak** [apache/airflow#29831] [General]
What will the best practices be to replace `prev_execution_date` or `DAG.previous_schedule()` after Airflow 3 release? I see they are now deprecated, but I cannot find an analogy.
Cannot use `DAG.get_last_dagrun().execution_date` because I need a prev schedule if there was no prev run.
It may be the wrong way to rely on these fields. Then please share your thoughts about the division of responsibility in this case.

Context: we're using `logical_date` and `prev_execution_date` to identify partitions to read between these dates. And if there were no runs, we don't want to do backfilling in the current dag run; we want to get the last schedule and read data only in these boundaries. 

Many thanks.

**Comment by potiuk** [comment]
cc: @uranusjr  might know the answer

**Reply by uranusjr** [reply]
Looks like it’s easiest to fetch the actual DAG run with `dag_run.get_previous_dagrun(state)` or `dag_run.get_previous_scheduled_dagrun()` (depending on how you define the last run).

**Reply by potiuk** [reply]
Just a note @uranusjr on that one (mental for me and for the future). We might want to think about adding macros serving similar cases for the internal API (AIP-44) case. We do not really want (and won't be able to)  to make a DB call to retrieve that information because the DagRun will be a dataclass rather than SQLAlchemy object and we will not have the DB session (cc @mhenc @pierrejeambrun @vincbeck ). This will be one of the (intended) limitations of the case where internal API will be used

**Reply by k-tomak** [reply]
@uranusjr, thank you for your answer. I've checked `dag_run.get_previous_dagrun(state)` and `dag_run.get_previous_scheduled_dagrun()`, and see by code that if we have no previous runs, it would be None. When we have the very first DAG run, we want to get here to the previous schedule, but not None. 
The main idea is to access even not the previous run but the previous schedule. This also helps not to make a DB call. So that's why`DAG.previous_schedule()` was very helpful, but now it's deprecated.

**Reply by uranusjr** [reply]
Sounds like your best choice would be to switch to a custom timetable and implement the method yourself then. The timetable object would be available via `dag.timetable`. I am hesitant to add back this kind of “peek back into history” functionalities since they only make sense for selected timetables, and have a history of being mis-used by people who are actually looking for real historic runs (for which `get_previous_scheduled_dagrun` is the correct choice).

We can always f the use case is prooved to be popular. (As far as I remember you’re the first asking for this, but timetables haven’t been around for that long so it may simply be a matter of time.)
